lib/libutils/ext/include/fault_mitigation.h

```diff
@@ -0,0 +1,687 @@
+
+/*
+ * struct ftmn_check - track current checked state
+ * @steps:	accumulated checkpoints
+ * @res:	last stored result or return value
+ *
+ * While a function is executed it can update its state as a way of keeping
+ * track of important passages inside the function. Before the function
+ * returns with for instance ftmn_return_res() to check that the
+ * accumulated state matches the expected state.
+ *
+ * @res is xored with FTMN_DEFAULT_HASH in order to retrieve the saved
+ * result or return value.
+ */
+struct ftmn_check {
+	unsigned long steps;
+	unsigned long res;
+};
+
+/*
+ * struct ftmn_func_arg - track a called function
+ * @hash:	xor bitmask
+ * @res:	stored result xored with @hash
+ *
+ * When the call of a function is tracked @hash is initialized to hash of
+ * caller xored with hash of called function. Before the called function
+ * updates @res it first xors @hash with its own hash, which is supposed to
+ * restore @hash to the hash of the calling function. This allows the
+ * calling function to confirm that the correct function has been called.
+ */
+struct ftmn_func_arg {
+	unsigned long hash;
+	unsigned long res;
+};
+
+/*
+ * struct ftmn - link a tracked call chain
+ * @check:	local checked state
+ * @arg:	argument for the next called tracked function
+ * @saved_arg:	pointer to an optional argument passed to this function
+ * @arg_pp:	cached return value from __ftmn_get_tsd_func_arg_pp()
+ * @my_hash:	the hash of the calling function
+ * @called_hash:the hash of the called function
+ *
+ * In order to maintain the linked call chain of tracked functions the
+ * struct ftmn_func_arg passed to this function is saved in @saved_arg
+ * before updating the argument pointer with @arg.
+ */
+struct ftmn {
+	struct ftmn_check check;
+	struct ftmn_func_arg arg;
+	struct ftmn_func_arg *saved_arg;
+	struct ftmn_func_arg **arg_pp;
+	unsigned long my_hash;
+	unsigned long called_hash;
+};
+
+/*
+ * enum ftmn_incr - increase counter values
+ *
+ * Prime numbers to be used when increasing the accumulated state.
+ * Different increase counters can be used to keep apart different
+ * checkpoints.
+ */
+enum ftmn_incr {
+	FTMN_INCR0 = 7873,
+	FTMN_INCR1 = 7877,
+	FTMN_INCR2 = 7879,
+	FTMN_INCR3 = 7883,
+	FTMN_INCR4 = 7901,
+	FTMN_INCR5 = 7907,
+	FTMN_INCR_RESERVED = 7919,
+};
+
+typedef int (*ftmn_memcmp_t)(const void *p1, const void *p2, size_t nb);
+
+/* The default hash used when xoring the result in struct ftmn_check */
+#ifdef __ILP32__
+#define FTMN_DEFAULT_HASH	0x9c478bf6UL
+#else
+#define FTMN_DEFAULT_HASH	0xc478bf63e9500cb5UL
+#endif
+
+/*
+ * FTMN_PANIC() - FTMN specific panic function
+ *
+ * This function is called whenever the FTMN function detects an
+ * inconsistency.  An inconsistency is able to occur if the system is
+ * subject to an fault injection attack, in this case doing a panic() isn't
+ * an extreme measure.
+ */
+#ifdef __KERNEL__
+#define FTMN_PANIC()	panic();
+#else
+#define FTMN_PANIC()	TEE_Panic(0);
```

Comment:
`TEE_ERROR_SECURITY` or `TEE_ERROR_BAD_STATE` could be good candidates.
But i admit it will not be specifically obvious it specifically from the fault injection countermeasure.

Reply:
The only occasion when I have found specific values to be helpful in `TEE_Panic()`, is when running negative tests in CI, when the `xtest` output is visible but not the secure world log. Not very applicable to this case I suppose. But FWIW either ways are OK with me (0 or something else).

Reply:
I don't care which we take either. @etienne-lms which do you prefer?

Reply:
Not willing to track glitch attacks on CI runs? :)
Ok, `0` is fine.